src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,49 +1,109 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int ARRAY_SIZE = 10;
+    private static final double INCREASE_ARRAY = 1.5;
+    private Object[] arraylist;
+    private int size;
+
+    public ArrayList() {
+        arraylist = new Object[ARRAY_SIZE];
+        size = 0;
+    }
 
     @Override
     public void add(T value) {
-
+        if (size == arraylist.length) {
+            grow();
+        }
+        arraylist[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        indexCheckForAdd(index);
+        if (size == arraylist.length) {
+            grow();
+        }
+        System.arraycopy(arraylist, index, arraylist, index + 1, size - index);
+        arraylist[index] = value;
+        size++;
+    }
 
+    private void grow() {
+        Object[] newArray = new Object[(int) (arraylist.length * INCREASE_ARRAY)];
+        System.arraycopy(arraylist,0,newArray,0,size);
+        arraylist = newArray;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        if (!list.isEmpty() && (size() + list.size() > arraylist.length)) {
+            grow();
+        }
+        for (int j = 0; j < list.size(); j++) {
+            add(list.get(j));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        indexCheck(index);
+        return (T) arraylist[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        indexCheck(index);
+        arraylist[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        indexCheck(index);
+        T returnValue = (T) arraylist[index];
+        System.arraycopy(arraylist, index + 1, arraylist, index, size - index - 1);
+        size--;
+        return returnValue;
     }
 
     @Override
     public T remove(T t) {
-        return null;
+        int index = getElementIndex(t);
+        return remove(index);
+    }
+
+    private int getElementIndex(T t) {
+        for (int i = 0; i < size(); i++) {
```

Comment:
don't call `size()` when we have direct access to `size` variable

Reply:
done